### ISSUE TITLE
k8s: Consensus affinity

### ISSUE THREAD

**Issue by jrmanes** [celestiaorg/devops#231]
## Description 📜

ℹ️  *Low priority, nice to have* 😌 

Hello team,

We realised during a Kubernetes upgrade that for minimising the downtime, it would be great to have also the consensus services in their own node pools (called: infra).

### Why?

Because right now we have two different node pools, one for DA and other for infra, the DA services are okay, they are placed only in their DA node pool, but for the consensus is different, they can be places anywhere.

### Issue

The node pools during the upgrade are being in parallel, we already set the max_surge to 1, meaning that we will only upgrade the nodes one by one, but, here is the problem, if we have the consensus deployed one in the DA node pools and the other in the the infra one, we could have a downtime.

### How to solve it?

We can use the pod affinity in the consensus services in order to don't place the nodes in the DA and, do not reuse one node twice.

This will guarantee us that all the consensus nodes are in their own nodes, and we won't have any issues during the upgrades.

If you have more in mind, please, feel free to add them to the list 📝 

Thanks in advance!

--- 

### Some interesting articles 🔍

[run-all-pods-in-single-node](https://stackoverflow.com/questions/37784480/avoiding-kubernetes-scheduler-to-run-all-pods-in-single-node-of-kubernetes-clust)
[scheduler](https://stackoverflow.com/questions/37784480/avoiding-kubernetes-scheduler-to-run-all-pods-in-single-node-of-kubernetes-clust)

--- 

### Tasks 🔖

- [x] Add the node affinity to the consensus
- [x] Verify that we don't schedule the consensus services in the DA node pool
- [x] Add the affinity to don't place the consensus pods twice in a node

--- 

### Repository 🗂️

- [celestia-stack-internal](https://github.com/celestiaorg/celestia-stack-internal)

---

#### Issues 👨‍🔧

Closes: 

--- 

PS: *Please, feel free to update any part of the ticket, any suggestion/idea is always welcome* 😊 

---

Best 🚀 

Jose Ramon Mañes

**Comment by smuu** [comment]
Is this still needed now that we know that: https://github.com/celestiaorg/devops/issues/198#issuecomment-1514967225

**Comment by jrmanes** [comment]
> Is this still needed now that we know that: [#198 (comment)](https://github.com/celestiaorg/devops/issues/198#issuecomment-1514967225)

I think yes, at least the part to don't schedule two consensus-validators in the same node, in order to prevent any downtime if we have two of them in the same node

**Comment by smuu** [comment]
> > Is this still needed now that we know that: [#198 (comment)](https://github.com/celestiaorg/devops/issues/198#issuecomment-1514967225)
> 
> I think yes, at least the part to don't schedule two consensus-validators in the same node, in order to prevent any downtime if we have two of them in the same node

Okay, about the part with the same node, I agree :+1:

**Comment by jrmanes** [comment]
the following code adds the affinity to the consensus node-pool

it does the actions mentioned, only in the `nodepool: pool-infra-0` & only one per node to prevent any downtime


```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app: consensus
  name: consensus
spec:
  template:
    spec:
      containers:
      - name: consensus
      affinity:
        nodeAffinity: # deploy only to the specific node-pool
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: k8s.scaleway.com/pool-name
                operator: In
                values:
                - pool-infra-0
        podAntiAffinity: # deploy only once per node
          requiredDuringSchedulingIgnoredDuringExecution:
          - topologyKey: kubernetes.io/hostname
            labelSelector:
              matchLabels:
                app: consensus
```

**Comment by jrmanes** [comment]
already tested 🆗 
deployed:
![Screenshot 2023-05-15 at 16 48 33](https://github.com/celestiaorg/devops/assets/32740567/a5d347bb-ddb8-49a6-b4bb-f87ba903c996)

redeployed to be sure:
![Screenshot 2023-05-15 at 16 50 43](https://github.com/celestiaorg/devops/assets/32740567/ba6adad2-f4b1-42fe-be3d-6838a1bc80b0)
